smart-wallets/n-key-wallet/n-key-wallet.pact

```diff
@@ -0,0 +1,106 @@
+(define-keyset 'admin-keyset (read-keyset 'admin-keyset))
+
+(module n-key-wallet 'admin-keyset
+  @doc"'n-key-wallet' provides safe transactions of Kadena coins.                \
+  \ The accounts in the wallet are guarded by a list of keys and a guardian key. \
+  \ Each transfer will require n number of keys to be signed depending on the    \
+  \ ratio of (amount to transfer) : (total balance). For example, if an account  \
+  \ has a balance of 10.0 coins and guard list with 2 keys, then the transfer    \
+  \ of 0-5 coins will require 1 key to be signed, and the transfer of 5-10 coins \
+  \ will require 2 keys to be signed. Guardian key will allow the rotation of    \
+  \ guard list in case some keys are lost."
+
+  (defschema user
+    user:string
+    guards:list
+    guardian:guard)
+
+  (deftable user-table:{user})
+
+  (defun smart-guard:bool ()
+    @doc "Module guard"
+    (create-module-guard 'wallet-guard)
+    )
+
+  (defun create-smart-user (user:string guards:[guard] guardian:guard)
+    @doc "Create a coin account managed by customized smart wallet"
+    (enforce-guard guardian)
+    (insert user-table user {
+      "user": user,
+      "guards": guards,
+      "guardian": guardian
+      })
+    (coin.create-account user (smart-guard))
+  )
+
+  (defun smart-transfer (account receiver amount)
+    @doc "Runs coin.transfer with customized rules."
+    (enforce-signed account amount)
+    (coin.transfer account receiver amount))
+
+  (defun rotate-guards (user:string guards:[guard])
+    @doc "Rotate guards of the user"
+    (enforce-guardian user)
+    (update user-table user {
+      "guards": guards
+      }))
```

Comment:
We probably want to scope this with a managed cap if possible

Reply:
Agreed, the defun `enforce-signed` could actually be a defcap `(SMART_TRANSFER account receiver amount)` which would then protect the signatures in the transaction. Managed is also easier now as for one-shot management you can simply use a bare `@managed` tag. Without it we're back to the signature-leak problem, which granted is only an issue in third-party txs. Also see my comment about fungible.

Reply:
However this would just be `TRANSFER` if you decide it makes sense to make this wallet a `fungible`.